feat: add `stats/incr/nanewvariance`

Adithya-jh opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves #5558 .

## Description

> What is the purpose of this pull request?

This pull request:

-   This pull request adds the nanewvariance module, which computes an incremental exponentially weighted variance while ignoring NaN values in the input data.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5558 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Adithya-jh committed:
test(stats/incr): update test cases for nanewvariance3

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Adithya-jh commented:
Hi  @kgryte 
I have implemented the nanewvariance module. could you please review my code and tell me if there are any changes to be made.

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanewvariance/lib/main.js:
```diff
@@ -0,0 +1,81 @@
+// MAIN //
+
+/**
+* Returns an accumulator function which incrementally computes an arithmetic mean, ignoring `NaN` values.
```
JSDoc says "arithmetic mean" — should be "exponentially weighted variance". Same issue in `lib/index.js` (line 22).

```suggestion
* Returns an accumulator function which incrementally computes an exponentially weighted variance, ignoring `NaN` values.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanewvariance/lib/main.js:
```diff
@@ -0,0 +1,81 @@
+/**
+* @license Apache-2.0
+*
+* Copyright (c) 2018 The Stdlib Authors.
```
Copyright year should be 2025, not 2018.

```suggestion
* Copyright (c) 2025 The Stdlib Authors.
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanewvariance/package.json:
```diff
@@ -0,0 +1,66 @@
+{
+  "name": "@stdlib/stats/incr/nanewvariance",
+  "version": "0.0.0",
+  "description": "Compute an exponentially weighted variance incrementally, ignoring `NaN` values.",
```
Use plain `NaN` (no backticks) in `package.json` descriptions.

```suggestion
  "description": "Compute an exponentially weighted variance incrementally, ignoring NaN values.",
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanewvariance/docs/types/index.d.ts:
```diff
@@ -0,0 +1,77 @@
+*
+* @param alpha - smoothing factor
+* @throws must provide a nonnegative number
+* @throws must be on the interval `[0,1]`
+* @returns accumulator function
+*
+* @example
+* var accumulator = nanewvariance( 0.5 );
+*
+* var v = accumulator();
+* // returns null
+*
+* v = accumulator( 2.0 );
+* // returns null
+*
+* v = accumulator( 4.0 );
+* // returns 2.0
+*
+* v = accumulator( NaN );
+* // returns 2.0
+*
+* v = accumulator( 5.0 );
+* // returns 2.3
+*
+* v = accumulator();
+* // returns 2.3
+*/
+declare namespace incrnanewvariance {
+	/**
+	* If provided a value, returns an updated exponentially weighted variance; otherwise, returns the current variance.
+	*
+	* @param x - input value
+	* @returns exponentially weighted variance or null
+	*/
+	type Accumulator = ( x?: number ) => number | null;
+}
+
+// EXPORTS //
+
+export = incrnanewvariance;
```
Missing `declare function` — without it, the exported namespace isn't callable and TypeScript consumers can't use the package. Also, the standalone `Accumulator` type at the top (lines 23–33) is unused.

Follow the pattern in `stats/incr/ewvariance`:

```ts
type accumulator = ( x?: number ) => number | null;

declare function incrnanewvariance( alpha: number ): accumulator;

export = incrnanewvariance;
```

Planeshifter commented on lib/node_modules/@stdlib/stats/incr/nanewvariance/examples/index.js:
```diff
@@ -0,0 +1,39 @@
+for ( i = 0; i < 100; i++ ) {
+	v = (randu() < 0.2) ? NaN : randu() * 100.0;
+	s2 = accumulator( v );
+	console.log('%s\t%s', ((isNaN(v)) ? 'NaN' : v.toFixed(4)), ((s2 === null) ? 'null' : s2.toFixed(4)));
```
Use `@stdlib/math/base/assert/is-nan` instead of the built-in `isNaN`. Also, stdlib style requires spaces inside parentheses: `isnan( v )`, not `isNaN(v)`.

Same spacing issue with `((s2 === null)` — should be `( s2 === null )`.

Planeshifter reviewed (changes requested):
Some late feedback in case there is appetite to move this PR along. Thanks for your work!

Planeshifter committed:
fix: address review feedback

Correct JSDoc and package descriptions to describe an exponentially
weighted variance, add a callable `declare function` signature and
remove the unused standalone type in the TypeScript declarations, and
use `@stdlib/math/base/assert/is-nan` with stdlib spacing conventions
in the example.

Planeshifter commented:
/stdlib merge

Planeshifter commented:
/stdlib update-copyright-years

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feature/nanewvariance

Planeshifter committed:
chore: update copyright years

Planeshifter committed:
refactor: rename accumulator factory to incrnanewvariance

Aligns the function and module export with the incr-prefixed naming convention used across stats/incr packages and the package's own docs/types.

Planeshifter committed:
style: use format for benchmark name

Replace string concatenation with format( '%s::accumulator', pkg ) and add the @stdlib/string/format require.

Planeshifter committed:
docs: fix equation annotations and rename to incrnanewvariance

Correct the equation label/raw text and SVG image source, and rename README references to the incr-prefixed identifier.

Planeshifter committed:
style: add blank line before test section separator

Match the two-blank-line convention preceding the `// TESTS //`
separator used across sibling packages.

Planeshifter reviewed (commented):
Looks good to me after clean-up. Thank you, @Adithya-jh!

kgryte commented:
The titles of these PRs need to be cleaned up.
